osu.Game.Rulesets.Taiko/Difficulty/Preprocessing/Colour/Data/CoupledColourEncoding.cs

```diff
@@ -0,0 +1,89 @@
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Collections.Generic;
+
+namespace osu.Game.Rulesets.Taiko.Difficulty.Preprocessing.Colour.Data
+{
+    /// <summary>
+    /// Encodes a list of <see cref="ColourEncoding"/>s, grouped together by back and forth repetition of the same
+    /// <see cref="ColourEncoding"/>. Also stores the repetition interval between this and the previous <see cref="CoupledColourEncoding"/>.
+    /// </summary>
+    public class CoupledColourEncoding
+    {
+        /// <summary>
+        /// Maximum amount of <see cref="CoupledColourEncoding"/>s to look back to find a repetition.
+        /// </summary>
+        private const int max_repetition_interval = 16;
+
+        /// <summary>
+        /// The <see cref="ColourEncoding"/>s that are grouped together within this <see cref="CoupledColourEncoding"/>.
+        /// </summary>
+        public readonly List<ColourEncoding> Payload = new List<ColourEncoding>();
+
+        /// <summary>
+        /// The previous <see cref="CoupledColourEncoding"/>. This is used to determine the repetition interval.
+        /// </summary>
+        public readonly CoupledColourEncoding? Previous;
+
+        /// <summary>
+        /// How many <see cref="CoupledColourEncoding"/> between the current and previous identical <see cref="CoupledColourEncoding"/>.
+        /// If no repetition is found this will have a value of <see cref="max_repetition_interval"/> + 1.
+        /// </summary>
+        public int RepetitionInterval { get; private set; } = max_repetition_interval + 1;
+
+        public CoupledColourEncoding(CoupledColourEncoding? previous)
+        {
+            Previous = previous;
+        }
+
+        /// <summary>
+        /// Returns true if other is considered a repetition of this encoding. This is true if other's first two payloads
+        /// have identical mono lengths.
+        /// </summary>
+        private bool isRepetitionOf(CoupledColourEncoding other)
+        {
+            if (Payload.Count != other.Payload.Count) return false;
+
+            for (int i = 0; i < Math.Min(Payload.Count, 2); i++)
+            {
+                if (!Payload[i].HasIdenticalMonoLength(other.Payload[i])) return false;
+            }
+
+            return true;
+        }
+
+        /// <summary>
+        /// Finds the closest previous <see cref="CoupledColourEncoding"/> that has the identical <see cref="Payload"/>.
+        /// Interval is defined as the amount of <see cref="CoupledColourEncoding"/> chunks between the current and repeated encoding.
+        /// </summary>
+        public void FindRepetitionInterval()
+        {
+            if (Previous?.Previous == null)
+            {
+                RepetitionInterval = max_repetition_interval + 1;
+                return;
+            }
+
+            CoupledColourEncoding? other = Previous.Previous;
```

Comment:
Why is this going back two steps, i.e. starting from `Previous.Previous` rather than... you know, just `Previous`?

Reply:
Missed this, this was originally because when `CoupledColourEncoding.isRepetitionOf` was a strict comparison, it's impossible for `Previous` to be a repetition of the current `CoupledColourEncoding`, as it will just be encoded within the same chunk. With the more lenient definition of repetition it seems like it should be possible to have `Previous` as a repetition.

Locally when I changed this it doesn't really change values in the set of maps I've tested, since it's a pretty rare case. But it might be a good idea to have some more testing (I will discuss that with the committee).